### PR TITLE
Thumbnail generation

### PR DESCRIPTION
**Pull request by geek1011**
Reworks the thumbnail generation to be faster and better match Kobo's behavior.

See #16 and #9.

### CODE REVIEW

File: kobo-uncaged/kutypes.go [hunk 1]
```diff
@@ -167,3 +167,26 @@ func (d koboDevice) CoverSize() (fullCover, libFull, libGrid image.Point) {
...
+// See https://github.com/shermp/Kobo-UNCaGED/issues/16#issuecomment-494229994
+// for more details.
+func (d koboDevice) FullCover() image.Point {
+	switch d {
+	case auraOne, auraOneLE: // daylight
+		return image.Pt(1404, 1872)
+	case gloHD, claraHD: // alyssum, nova
+		return image.Pt(1072, 1448)
+	case auraHD, auraH2O, auraH2Oed2r1, auraH2Oed2r2: // dragon
+		return image.Pt(1080, 1440)
```
Comment (NiLuJe) [inline kobo-uncaged/kutypes.go]: The H2O's effective resolution is really 1080x1429, Nickel is doing it wrong ;).

This doesn't apply to the H2O², though (AFAIK).

File: kobo-uncaged/util.go [hunk 2]
```diff
@@ -74,3 +75,15 @@ func readJSON(fn string, out interface{}) (emptyOrNotExist bool, err error) {
...
 }
+
+// resizeKeepAspectRatioByExpanding resizes a sz to fill bounds while keeping
+// the aspect ratio. It is based on Qt::KeepAspectRatioByExpanding.
+func resizeKeepAspectRatioByExpanding(sz image.Point, bounds image.Point) image.Point {
+	if sz.X == 0 || sz.Y == 0 {
+		return sz
+	}
+	if rw := bounds.Y * sz.X / sz.Y; rw >= bounds.X {
+		return image.Pt(rw, bounds.Y)
```
Comment (NiLuJe) [inline kobo-uncaged/util.go]: Not familiar with Go, but this technically ought to be round'ed instead of truncated (AFAICT, Qt truncates, while IM rounds).

Comment (NiLuJe) [reply kobo-uncaged/util.go]: FWIW, that's where the mismatch between the 149x199 I got with my text script and the 149x198 Nickel generates ;).

Comment (NiLuJe) [reply kobo-uncaged/util.go]: This is highly nitpicky, so if you prefer to honor Qt's behavior, that's perfectly fine by me ;).

Comment (shermp) [reply kobo-uncaged/util.go]: Point uses integers, so yeah, that will truncate.

Comment (geek1011) [reply kobo-uncaged/util.go]: Yep, it truncates. I'm following Qt's behavior (I verified the tests with Qt's resizing).

File: kobo-uncaged/kutypes.go (hunk 1 above)
Comment (geek1011) [reply kobo-uncaged/kutypes.go]: I think I'll use that then.

File: kobo-uncaged/main.go [hunk 3]
```diff
@@ -502,60 +479,42 @@ func (ku *KoboUncaged) saveDeviceInfo() error {
...
 
-	var libFW, libFH, gridFW, gridFH int
-	if (float64(size.X) / float64(size.Y)) < 1.0 {
-		libFH, gridFH = libFullSize.Y, libGridSize.Y
-	} else {
-		libFW, gridFW = libFullSize.X, libGridSize.X
-	}
+	for _, cover := range []koboCover{libFull, libGrid} {
+		nsz := resizeKeepAspectRatioByExpanding(sz, cover.Size(ku.device))
+		nimg := imaging.Resize(img, nsz.X, nsz.Y, imaging.Linear)
```
Comment (geek1011) [inline kobo-uncaged/main.go]: Which image resize library/algorithm are we planning to use?

Comment (shermp) [reply kobo-uncaged/main.go]: We may have to upscale the full cover, so an algorithm that's only decent for downscaling probably isn't the best idea.

Rez seemed pretty quick, so I would be happy with that. Bilinear or Bicubic perhaps. Or we could make the algorithm selectable in the config file, and the quality purists among us can select a better algorithm if they want to take the performance hit.

File: kobo-uncaged/util.go [hunk 4]
```diff
@@ -87,3 +87,17 @@ func resizeKeepAspectRatioByExpanding(sz image.Point, bounds image.Point) image.
...
+// .kobo-images/{dir1}/{dir2}/{basename} - N3_SOMETHING.jpg
+func hashedImageParts(imageID string) (dir1, dir2, basename string) {
+	imgID := []byte(imageID)
+	h := uint32(0x00000000)
+	for _, x := range imgID {
+		h = (h << 4) + uint32(x)
+		h ^= (h & 0xf0000000) >> 23
+		h &= 0x0fffffff
+	}
+	return dir1, dir2, imageID
```
Comment (shermp) [inline kobo-uncaged/util.go]: Umm.. you appear to be missing some code here to actually generate the outputs?

Comment (geek1011) [reply kobo-uncaged/util.go]: I fixed that just now. I'm fixing the relpath too (I mixed up .parsed and .jpg).

**Comment by geek1011** [comment]
This is ready for testing (and choosing the resize library/algorithm).

File: kobo-uncaged/kutypes.go [hunk 5]
```diff
@@ -195,3 +156,43 @@ func (d koboDevice) FullCover() image.Point {
...
+func (k koboCover) String() string {
+	switch k {
+	case fullCover:
+		return "N3_FULL"
+	case libFull:
+		return "N3_LIBRARY_FULL"
+	case libGrid:
+		return "N3_LIBRARY_GRID"
+	default:
+		panic("unknown cover type")
```
Comment (shermp) [inline kobo-uncaged/kutypes.go]: Just a heads up that I'm generally not a fan of panicing. Especially since we're being run from a rather delicate shell script, where its nice to have a known return value.

In this case though, I guess its fine. There should be no reason this ever occurs, except for if we make a typo in the code somewhere.

Comment (geek1011) [reply kobo-uncaged/kutypes.go]: That's the same way I am about it. I only panic if it is an issue with the function's own code. I never panic if it could be triggered by an external library, if it is a runtime error, or in the case of a library if it could be triggered by the user. The only exception I make is if it is a one-off command or if it is dangerous to continue running.

Comment (shermp) [reply kobo-uncaged/kutypes.go]: That makes sense.

**Comment by shermp** [comment]
Just added a small update to request the full size cover from Calibre, so we aren't upscaling a tiny thumbnail!

**Comment by shermp** [comment]
Nickel still appears to be generating its own thumbnails for me. It is very obvious when sending larger numbers of books at a time (eg: > 5 books). It may appear to work when sending one or two books, but I don't think it actually is.

I'm going out in a few minutes, but I'll try and do some more investigation when I get back later.

**Comment by NiLuJe** [comment]
Yep, confirmed. I'm not getting *anything* (thumbnails-wise) getting generated by KU.

It's easy enough to check by SSH'ing into the device while it's still connected to Calibre.
With a `find /mnt/newonboard/.kobo-images/ -name '*BLAH*'`, with BLAH being a relevant bit of the book's filename, I used to see the small thumbnails freshly created ;).

Now, zilch. And Nickel indeed generates all three in one go the first time you open the Library (or automatically if one pops up at the front of one of the Home tiles, generally, My Books).

**Comment by geek1011** [comment]
Ok, I'll test this myself once I get Calibre set up. I've been working on this blindly so far.

**Comment by geek1011** [comment]
@NiLuJe @shermp I think it should work fine now.

**Comment by NiLuJe** [comment]
Oh, right, found a few of 'em in `/tmp/ku/.kobo-images` :D.

The set isn't complete, though, sometimes I have a `LIBRARY_FULL`, sometimes both `LIBRARY_FULL` and `LIBRARY_GRID`...

Will check again w/ the latest commit ;).

**Comment by NiLuJe** [comment]
Okay, yeah, I'm only ever getting KU to generate `LIBRARY_FULL` & `LIBRARY_GRID`, and it takes quite a bit of time to scale 'em, without any specific on-screen feedback (I had to check top to confirm that it was pegged at 100% CPU after being puzzled at not seeing them pop-up immediately, but a few dozen seconds later ;)).

**Comment by NiLuJe** [comment]
I'm also getting slightly wonky thumbnail sizes:

Given a Calibre cover.jpg @ 1391x2200, I'd expect GRID to be 149x236, but it's 149x352; and LIB_FULL to be 355x561, but it's 355x838

This is possibly a side-effect of what Calibre sends as a source?

**Comment by shermp** [comment]
Ok, can see why the full cover isn't been created. Just need to add it to the array at `for _, cover := range []koboCover{libFull, libGrid}`

Not sure about the thumbnail sizes, but it looks to me like the wrong axis is being scaled. Calibre sends a base64 string, and a separate width and height.

**Comment by geek1011** [comment]
> I'm only ever getting KU to generate LIBRARY_FULL & LIBRARY_GRID

That's all it's currently configured to generate. Should I add `FULL` to the list of images to generate?

> it takes quite a bit of time to scale 'em

The algorithm has not been changed yet.

> I'm also getting slightly wonky thumbnail sizes ... This is possibly a side-effect of what Calibre sends as a source?

Possibly. Can you check what Qt does with them?

**Comment by shermp** [comment]
> That's all it's currently configured to generate. Should I add `FULL` to the list of images to generate?

Yes, because as @NiLuJe discovered, and also confirmed by @davidfor we need all three otherwise Nickel will scale the tiny thumbnail

**Comment by NiLuJe** [comment]
The number I pulled were from my quck'n dirty Python port of Qt's QSize:scaled, but based on the *original* cover resolution, which is potentially not what's actually happening here ;).

**Comment by NiLuJe** [comment]
Yep, after a bit of extra logging, `resizeKeepAspectRatioByExpanding` is doing something wrong.

Note that the full cover should also *not* be using `Qt::KeepAspectRatioByExpanding`, but `Qt::KeepAspectRatio` instead ;).

**Comment by geek1011** [comment]
> Yep, after a bit of extra logging, resizeKeepAspectRatioByExpanding is doing something wrong.

Could you provide some test data for it?

> Note that the full cover should also not be using Qt::KeepAspectRatioByExpanding, but Qt::KeepAspectRatio instead ;).

Good catch. I'll implement that next.

File: kobo-uncaged/util.go [hunk 6]
```diff
@@ -67,3 +69,29 @@ func readJSON(fn string, out interface{}) (emptyOrNotExist bool, err error) {
...
+// resizeKeepAspectRatioByExpanding resizes a sz to fill bounds while keeping
+// the aspect ratio. It is based on Qt::KeepAspectRatioByExpanding.
+func resizeKeepAspectRatioByExpanding(sz image.Point, bounds image.Point) image.Point {
+	if sz.X == 0 || sz.Y == 0 {
+		return sz
+	}
+	if rw := bounds.Y * sz.X / sz.Y; rw >= bounds.X {
+		return image.Pt(rw, bounds.Y)
+	}
+	return image.Pt(bounds.X, bounds.Y*sz.Y/sz.X)
```
Comment (NiLuJe) [inline kobo-uncaged/util.go]: Should be `bounds.X*sz.Y/sz.X` ;)

**Comment by shermp** [comment]
On the full cover, I just had a thought that I might make generating it a configurable option.

Point of that is, there are probably others, like me, who disable showing the cover on the sleep screen, so the full cover isn't used (unless it's used else where that I don't know of). Not using the full cover means requesting a smaller image from Calibre, which means a smaller transfer, and quicker downscaling.

Thoughts?

File: kobo-uncaged/util.go (hunk 6 above)
Comment (geek1011) [reply kobo-uncaged/util.go]: Fixed, but my tests still pass. I think my tests are missing some edge cases. I'm probably going to write something to generate some new ones automatically so I don't miss cases.

**Comment by NiLuJe** [comment]
If Point is an int, doesn't the division need to explicit cast to a float so we get anything useful out of it, instead of 1 or 0?

**Comment by geek1011** [comment]
@NiLuJe Yes. TBH, I haven't actually got my head completely wrapped around the logic behind the Qt equations (and now I realize that my unit tests didn't include fractional ratios).

**Comment by NiLuJe** [comment]
@geek1011 : This simple change made it easier to grok on my end:

```diff
index 6f9d37a..ddbc894 100644
--- a/kobo-uncaged/util.go
+++ b/kobo-uncaged/util.go
@@ -79,7 +79,8 @@ func resizeKeepAspectRatio(sz image.Point, bounds image.Point, expand bool) imag
        }
 
        var useHeight bool
-       rw := int(float64(bounds.Y) * float64(sz.X) / float64(sz.Y))
+       ar := float64(sz.X) / float64(sz.Y)
+       rw := int(float64(bounds.Y) * ar)
 
        if !expand {
                useHeight = rw <= bounds.X
@@ -90,7 +91,7 @@ func resizeKeepAspectRatio(sz image.Point, bounds image.Point, expand bool) imag
        if useHeight {
                return image.Pt(rw, bounds.Y)
        }
-       return image.Pt(bounds.X, int(float64(bounds.X)*float64(sz.Y)/float64(sz.X)))
+       return image.Pt(bounds.X, int(float64(bounds.X) / ar))
 }
 
 // hashedImageParts returns the parts needed for constructing the path to the
```

**Comment by NiLuJe** [comment]
As for my bit of logging:

```
diff --git a/kobo-uncaged/main.go b/kobo-uncaged/main.go
index 365ab7e..05c4563 100644
--- a/kobo-uncaged/main.go
+++ b/kobo-uncaged/main.go
@@ -498,6 +498,7 @@ func (ku *KoboUncaged) saveCoverImage(contentID string, size image.Point, imgB64
        for _, cover := range []koboCover{fullCover, libFull, libGrid} {
                nsz := cover.Resize(ku.device, sz)
                nimg := imaging.Resize(img, nsz.X, nsz.Y, imaging.Linear)
+               log.Printf("Resized %dx%d cover to %dx%d -> %dx%d => %dx%d\n", sz.X, sz.Y, cover.Size(ku.device).X, cover.Size(ku.device).Y, nsz.X, nsz.Y, nimg.Bounds().Size().X, nimg.Bounds().Size().Y)
                nfn := filepath.Join(imgDir, cover.RelPath(imgID))
 
                if err := os.MkdirAll(filepath.Dir(nfn), 0755); err != nil {
```

And success:

```
May 22 05:36:37 KoboUNCaGED[2769]: 2019/05/22 05:36:37 Resized 903x1429 cover to 1080x1429 -> 903x1429 => 903x1429
May 22 05:37:05 KoboUNCaGED[2769]: 2019/05/22 05:37:05 Resized 903x1429 cover to 355x530 -> 355x561 => 355x561
May 22 05:37:23 KoboUNCaGED[2769]: 2019/05/22 05:37:23 Resized 903x1429 cover to 149x223 -> 149x235 => 149x235
```

Well, almost, because we could essentially skip the scaling pass for the full cover if sz == nsz ;).

File: kobo-uncaged/main.go [hunk 7]
```diff
@@ -497,10 +497,16 @@ func (ku *KoboUncaged) saveCoverImage(contentID string, size image.Point, imgB64
...
 		nfn := filepath.Join(imgDir, cover.RelPath(imgID))
 
-		log.Printf("Resized %s cover to %s (target %s) for %s\n", sz, nsz, cover.Size(ku.device), cover)
+		log.Printf("Resizing %s cover to %s (target %s) for %s\n", sz, nsz, cover.Size(ku.device), cover)
+
+		nimg := img
+		if !sz.Eq(nsz) {
+			nimg = imaging.Resize(nimg, nsz.X, nsz.Y, imaging.Linear)
+		} else {
+			log.Printf(" -- Skipped resize: already correct size")
```
Comment (NiLuJe) [inline kobo-uncaged/main.go]: -> `log.Println`?

Comment (NiLuJe) [reply kobo-uncaged/main.go]: Otherwise, good call on making it generic, that was on my notes for today ;).

Comment (geek1011) [reply kobo-uncaged/main.go]: I personally tend to prefer using printf for everything, as I find it easier to read (figuring out where I put newlines and where I didn't). But, for consistency, I'll change it here.

Comment (NiLuJe) [reply kobo-uncaged/main.go]: I agree with you, but the lack of a trailing LF confused me ;).

**Comment by shermp** [comment]
Fixed an issue with the WaitGroup. It actually does what it was supposed to now!

I REALLY wouldn't recommend sending large numbers of books at the moment. It is slooooow. Stick to three max.

**Comment by shermp** [comment]
@NiLuJe , @geek1011 if you are happy with this PR, I'm happy to merge it, and we can move on to performance optimizations.

**Comment by shermp** [comment]
I'll take the thumbs up as confirmation :)

Thank you both very much.
